pageserver/src/tenant/storage_layer.rs

```diff
@@ -200,56 +253,24 @@ impl ValuesReconstructState {
         &mut self,
         key: &Key,
         lsn: Lsn,
-        value: Value,
+        completes: bool,
+        value: sync::oneshot::Receiver<Result<OnDiskValue, std::io::Error>>,
     ) -> ValueReconstructSituation {
-        let state = self
-            .keys
-            .entry(*key)
-            .or_insert(Ok(VectoredValueReconstructState::default()));
-
-        if let Ok(state) = state {
-            let key_done = match state.situation {
-                ValueReconstructSituation::Complete => unreachable!(),
-                ValueReconstructSituation::Continue => match value {
-                    Value::Image(img) => {
-                        state.img = Some((lsn, img));
-                        true
-                    }
-                    Value::WalRecord(rec) => {
-                        debug_assert!(
-                            Some(lsn) > state.get_cached_lsn(),
-                            "Attempt to collect a record below cached LSN for walredo: {} < {}",
-                            lsn,
-                            state
-                                .get_cached_lsn()
-                                .expect("Assertion can only fire if a cached lsn is present")
-                        );
-
-                        let will_init = rec.will_init();
-                        state.records.push((lsn, rec));
-                        will_init
-                    }
-                },
-            };
+        let state = self.keys.entry(*key).or_default();
 
-            if key_done && state.situation == ValueReconstructSituation::Continue {
-                state.situation = ValueReconstructSituation::Complete;
-                self.keys_done.add_key(*key);
+        match state.situation {
+            ValueReconstructSituation::Complete => unreachable!(),
```

Comment:
nit: we can turn this into an `assert_eq!` and drop the condition for `push` and in the `completes` branch below.

Reply:
I'd rather keep it as is. I think the condition below reads easier this way (don't have to figure out control flow).

Reply:
Ok. I find it more confusing myself -- the `unreachable` here basically states that it is illegal to call this method on a `Complete` key, but then the condition below indicates that it is valid. If it's illegal to call this on `Complete` it seems more clear to state that via an assertion at the method entry point.

End result is the same though, so I won't press for it.